File: Convert Sorted List to Binary Search Tree/Convert_Sorted_List_to_Binary_Search_Tree.cpp
// Runtime: 44 ms, faster than 70.65% of C++ online submissions for Convert Sorted List to Binary Search Tree.
// Memory Usage: 30 MB, less than 0.87% of C++ online submissions for Convert Sorted List to Binary Search Tree.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution 
{
public:
    TreeNode* sortedListToBST(ListNode* head) 
    {
        if (head == 0)
            return 0;
        
        // 利用双指针找到链表的中点
        ListNode *slowPtr = head, *fastPtr = head, *prePtr = 0;
        while (fastPtr->next != 0 && fastPtr->next->next != 0)
        {
            prePtr = slowPtr;
            slowPtr = slowPtr->next;
            fastPtr = fastPtr->next->next;
        }

        // 建立根节点
        TreeNode *treeNode = new TreeNode(slowPtr->val);

        // 递归的建立左子树
        if (prePtr != 0)
        {
            prePtr->next = 0;
            treeNode->left = sortedListToBST(head);
            prePtr->next = slowPtr;
        }

        // 递归的建立右子树
        if (slowPtr->next != 0)
        {
            treeNode->right = sortedListToBST(slowPtr->next);
        }
        
        return treeNode;
    }
};

File: Convert Sorted List to Binary Search Tree/Convert_Sorted_List_to_Binary_Search_Tree.py
# Runtime: 136 ms, faster than 65.78% of Python3 online submissions for Convert Sorted List to Binary Search Tree.
# Memory Usage: 13.4 MB, less than 38.57% of Python3 online submissions for Convert Sorted List to Binary Search Tree.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedListToBST(self, head: 'ListNode') -> 'TreeNode':
        if head is None:
            return None
        slowPtr, fastPtr, prePtr = head, head, None
        while fastPtr.next is not None and fastPtr.next.next is not None:
            prePtr = slowPtr
            slowPtr = slowPtr.next
            fastPtr = fastPtr.next.next
        
        treeNode = TreeNode(slowPtr.val)
        
        if prePtr is not None:
            prePtr.next = None
            treeNode.left = self.sortedListToBST(head)
            prePtr.next = slowPtr

        if slowPtr.next is not None:
            treeNode.right = self.sortedListToBST(slowPtr.next)
        
        return treeNode